site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+					methods[methodName] = extractMethodMetadata(method, src)
+				}
+			}
+		}
+		return true
+	})
+```
+
+Let's define a couple of struct to help us collect information about methods:
+
+```go
+type methodParam struct {
+	Name string
+	Type string
+}
+
+type methodData struct {
+	Params        []methodParam
+	Results       []string
+}
+
+
+// for each found method, we'll store it's name, params with their types and return types in methods := map[string]methodData {}
+```
+
+Now let's write a short function to populate these structs with metadata about a method:
+
+```go
+func formatNode(src []byte, node ast.Expr) string {
+	return string(src[node.Pos()-1 : node.End()-1])
+}
+
+func extractMethodMetadata(method *ast.Field, src []byte) methodData {
+	params := []methodParam{}
+	results := []string{}
+	e := method.Type.(*ast.FuncType)
+	if e.Params != nil {
+		for _, param := range e.Params.List {
+			for _, paramName := range param.Names {
+				paramType := (formatNode(src, param.Type))
+				params = append(params, methodParam{Name: paramName.Name, Type: paramType})
+			}
+		}
+	}
+
+	if e.Results != nil {
+		for _, r := range e.Results.List {
+			typeStr := formatNode(src, r.Type)
+			if len(r.Names) > 0 {
+				for _, k := range r.Names {
+					results = append(results, fmt.Sprintf("%s %s", k.Name, typeStr))
+				}
+			} else {
+				results = append(results, typeStr)
+			}
+		}
+	}
+	return methodData{Params: params, Results: results}
+}
+```
+
+Now we can run the parser on our interface and we'll get something like: `map[Speak:{Params:[{Name:x Type:int}] Results:[]}]`
+As you can see, we collected all the information needed about interface methods and we can now move on to generating decorator with this data.
+
+### Code generation using templates
+
+Let's get to it! We'll start by defining a few helper functions that will be useful during code generation. They will operate on the metadata we've collected before.
+
+```go
+	helperFuncs := template.FuncMap{
+		"joinResults": func(results []string) string {
+			return strings.Join(results, ", ")
+		},
+		"joinResultsForSignature": func(results []string) string {
+			return fmt.Sprintf("(%s)", strings.Join(results, ", "))
+		},
+		"joinParams": func(params []methodParam) string {
+			paramsNames := []string{}
+			for _, param := range params {
+				s := param.Name
+				if strings.HasPrefix(param.Type, "...") {
+					s += "..."
+				}
+				paramsNames = append(paramsNames, s)
+			}
+			return strings.Join(paramsNames, ", ")
+		},
+		"joinParamsWithType": func(params []methodParam) string {
+			paramsWithType := []string{}
+			for _, param := range params {
+				paramsWithType = append(paramsWithType, fmt.Sprintf("%s %s", param.Name, param.Type))
+			}
+			return strings.Join(paramsWithType, ", ")
+		},
+	}
+```
+
+Next we'll create a [Go Template](https://golang.org/pkg/text/template/) for both our decorators:
+
+{{< highlight go "linenos=table" >}}
+	tracerTemplate := `
```

Comment:
Let's also add the code generation prefix template: `// Code generated by <>; DO NOT EDIT.`